File: src/test/java/com/carry_laundry/carry_backend/term/domain/entity/TermTest.java
package com.carry_laundry.carry_backend.term.domain.entity;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

import java.time.LocalDate;
import java.util.stream.Stream;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.Arguments;
import org.junit.jupiter.params.provider.MethodSource;

@DisplayName("Term 엔티티는")
class TermTest {

    private static final Long TERM_META_ID = 1L;
    private static final String CONTENT = "content";
    private static final Integer VERSION_COUNT = 1;
    private static final LocalDate CREATED_AT = LocalDate.of(2024, 12, 25);

    @Nested
    @DisplayName("생성할 때")
    class WhenCreated {

        @Test
        @DisplayName("정상적으로 생성된다.")
        void willCreated() {
            // Act
            Term actualResult = Term.of(TERM_META_ID, CONTENT, VERSION_COUNT, CREATED_AT);
            // Assert
            assertThat(actualResult).isNotNull()
                .extracting("termMetaId", "content", "versionCount", "createdAt")
                .containsExactly(TERM_META_ID, CONTENT, VERSION_COUNT, CREATED_AT);
        }

        @ParameterizedTest
        @MethodSource("blankStrings")
        @DisplayName("content가 null이거나 공백이면 예외가 발생한다.")
        void willThrowExceptionIfContentIsNullOrBlank(String expectedContent) {
            // Act & Assert
            assertThatThrownBy(
                () -> Term.of(TERM_META_ID, expectedContent, VERSION_COUNT, CREATED_AT))
                .isInstanceOf(IllegalArgumentException.class)
                .hasMessage("content should not be blank");
        }

        @Test
        @DisplayName("versionCount가 1보다 작으면 예외가 발생한다.")
        void willThrowExceptionIfVersionCountIsLessThanOne() {
            // Act & Assert
            assertThatThrownBy(
                () -> Term.of(TERM_META_ID, CONTENT, 0, CREATED_AT))
                .isInstanceOf(IllegalArgumentException.class)
                .hasMessage("versionCount should be greater than 0");
        }

        @Test
        @DisplayName("createdAt이 null이면 예외가 발생한다.")
        void willThrowExceptionIfCreatedAtIsNull() {
            // Act & Assert
            assertThatThrownBy(
                () -> Term.of(TERM_META_ID, CONTENT, VERSION_COUNT, null))
                .isInstanceOf(IllegalArgumentException.class)
                .hasMessage("createdAt should not be null");
        }

        private static Stream<Arguments> blankStrings() {
            return Stream.of(
                Arguments.of((String) null),
                Arguments.of(""),
                Arguments.of(" "),
                Arguments.of("  ")
            );
        }
    }

    @Test
    @DisplayName("getVersion 메서드는 생성일과 버전 카운트를 조합하여 반환한다.")
    void getVersion() {
        // Arrange
        Term term = Term.of(TERM_META_ID, CONTENT, VERSION_COUNT, CREATED_AT);
        // Act
        String actualResult = term.getVersion();
        // Assert
        assertThat(actualResult).isEqualTo(CREATED_AT + "_" + VERSION_COUNT);
    }

}

File: carry-operation/src/test/kotlin/com/carry/operation/domain/model/TermTest.kt
package com.carry.operation.domain.model

import com.carry.operation.domain.vo.TermType
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Nested
import org.junit.jupiter.api.Test
import java.time.Instant

class TermTest {

    @Nested
    inner class Create {

        @Test
        fun `약관을 생성하면 활성 상태이고 버전이 1이다`() {
            val term = Term.create(
                title = "서비스 이용약관",
                content = "약관 내용입니다.",
                type = TermType.SERVICE,
                required = true,
            )

            assertThat(term.id).isNull()
            assertThat(term.title).isEqualTo("서비스 이용약관")
            assertThat(term.content).isEqualTo("약관 내용입니다.")
            assertThat(term.type).isEqualTo(TermType.SERVICE)
            assertThat(term.required).isTrue()
            assertThat(term.version).isEqualTo(1)
            assertThat(term.active).isTrue()
            assertThat(term.createdAt).isNotNull()
            assertThat(term.updatedAt).isNotNull()
        }
    }

    @Nested
    inner class Update {

        @Test
        fun `약관을 수정하면 버전이 증가한다`() {
            val term = reconstitutedTerm()
            val previousVersion = term.version

            term.update(title = "수정된 제목", content = "수정된 내용", required = false)

            assertThat(term.version).isEqualTo(previousVersion + 1)
            assertThat(term.required).isFalse()
        }

        @Test
        fun `약관을 여러 번 수정하면 버전이 누적 증가한다`() {
            val term = reconstitutedTerm()

            term.update(title = "수정1", content = "내용1", required = true)
            term.update(title = "수정2", content = "내용2", required = false)

            assertThat(term.version).isEqualTo(3)
        }
    }

    @Nested
    inner class Deactivate {

        @Test
        fun `약관을 비활성화하면 active가 false가 된다`() {
            val term = reconstitutedTerm()

            term.deactivate()

            assertThat(term.active).isFalse()
        }

        @Test
        fun `비활성화된 약관을 활성화하면 active가 true가 된다`() {
            val term = reconstitutedTerm(active = false)

            term.activate()

            assertThat(term.active).isTrue()
        }
    }

    @Nested
    inner class VersionIncrement {

        @Test
        fun `수정할 때마다 updatedAt이 갱신된다`() {
            val term = reconstitutedTerm()
            val previousUpdatedAt = term.updatedAt

            term.update(title = "새 제목", content = "새 내용", required = true)

            assertThat(term.updatedAt).isAfterOrEqualTo(previousUpdatedAt)
        }
    }

    private fun reconstitutedTerm(
        active: Boolean = true,
    ) = Term.reconstitute(
        id = 1L,
        title = "서비스 이용약관",
        content = "약관 내용입니다.",
        type = TermType.SERVICE,
        required = true,
        version = 1,
        active = active,
        createdAt = Instant.now(),
        updatedAt = Instant.now(),
    )
}
